feat: migrate to quick-protobuf

kckeiks opened the pull request:
## Description

<!-- Please write a summary of your changes and why you made them.-->
<!-- This section will appear as the commit message after merging. Please craft it accordingly. -->
Instead of relying on `protoc` and buildscripts, we generate the bindings using `pb-rs` and version them within our codebase. This makes for a better IDE integration, a faster build and an easier use of `rust-libp2p` because we don't force the `protoc` dependency onto them.

Resolves #3024.

## Change checklist

<!-- Please add a Changelog entry in the appropriate crates and bump the crate versions if needed. See <https://github.com/libp2p/rust-libp2p/blob/master/docs/release.md#development-between-releases>-->

- [x] I have performed a self-review of my own code
- [x] I have made corresponding changes to the documentation
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] A changelog entry has been made in the appropriate crates

kckeiks committed:
Add generated directory

kckeiks committed:
Add generated code

kckeiks committed:
Add ci

kckeiks committed:
Try adding quotes

thomaseizinger commented on .github/workflows/ci.yml:
```diff
@@ -305,4 +305,22 @@ jobs:
+
+      - name: Generated proto files and diff
+        run: |
+          cargo install pb-rs 
```
Two things: Use `--locked` and if you add the cache action before this, we don't have to reinstall it every time.

thomaseizinger commented on .github/workflows/ci.yml:
```diff
@@ -305,4 +305,22 @@ jobs:
+        run: |
+          cargo install pb-rs 
+          pb-rs "**/generated/**.proto"
+          git diff --exit-code ${{ github.ref }}
```
Isn't the `github.ref` part redundant here? Git should automatically diff to that commit right?

thomaseizinger reviewed (approved):
Nice, this is what I was looking for, thanks!

Once we delete the other proto files, those should hopefully show up as moved rather than added.

mxinden commented on core/src/generated/mod.rs:
```diff
@@ -0,0 +1,4 @@
+// Automatically generated mod.rs
```
Note that we don't use `mod.rs` files anywhere else within this repository. Is this necessary to function or would a `generated.rs` file one level above work as well?

thomaseizinger replied:
```
❯ pb-rs --help
pb-rs 0.10.0
Johann Tuffe <tafia973@gmail.com>
A converter from proto files into quick-protobuf compatible Rust module

USAGE:
    pb-rs [FLAGS] [OPTIONS] [INPUT]...

FLAGS:
        --add-deprecated-fields    Add deprecated fields and mark them as #[deprecated]
    -e, --error-cycle              Error out if recursive messages do not have optional fields
    -D, --dont_use_cow             Don't use Cow for String and Byte types
        --gen-info                 Generate MessageInfo implementations
        --hashrown                 Use hashrown for HashMap implementation
        --nostd                    Generate no_std compliant code
    -H, --no-headers               Do not add module comments and module attributes in generated file
    -n, --no-output                Show enums and messages in this .proto file, including those imported. No code
                                   generated
        --owned                    Generate Owned structs when the proto stuct has a lifetime
    -s, --single-mod               Omit generation of modules for each package when there is only one package
    -h, --help                     Prints help information
    -V, --version                  Prints version information

OPTIONS:
    -R, --custom_repr <CUSTOM_REPR>                      The value to use for the optional #[repr(...)] for every struct
    -C, --custom_struct_derive <CUSTOM_STRUCT_DERIVE>
            The comma separated values to add to #[derive(...)] for every struct

    -I, --include <INCLUDE_PATH>                         Path to search for imported protobufs
    -o, --output <OUTPUT>
            Generated file name, defaults to INPUT with 'rs' extension, cannot be used with --output_directory

    -d, --output_directory <OUTPUT_DIR>
            Output directory of generated code, cannot be used with --output


ARGS:
    <INPUT>...    The .proto files used to generate quick-protobuf code
```

There doesn't seem to be an option to not generate a `mod.rs` file.

rkuhn replied:
FWIW: I find `mod.rs` more consistent than the libp2p approach, since it keeps all files of a module in the same place.

mxinden reviewed (commented):
Thanks to both of you for the work here!

kckeiks committed:
Add input variable

kckeiks committed:
Move inputs

kckeiks committed:
Move inputs 2

kckeiks committed:
Glob into variable

kckeiks committed:
Glob into variable 2

kckeiks committed:
Use glob

kckeiks committed:
Remove quotes

kckeiks committed:
Remove refs

kckeiks committed:
Test glob pattern

thomaseizinger committed:
Bad commit

thomaseizinger committed:
Remove explicit echo

This is unnecessary as we can see the command being executed in the
action log.

See https://github.com/libp2p/rust-libp2p/actions/runs/3898847742/jobs/6657969252#step:5:4.

thomaseizinger committed:
Fail CI on untracked files

thomaseizinger committed:
Revert "Bad commit"

This reverts commit 391a892a8a1b30105493580fbd1afd3dd68f091e.

thomaseizinger committed:
Cache installed `pb-rs` binary

thomaseizinger committed:
Show bad proto files to the user

thomaseizinger committed:
Don't duplicate command

thomaseizinger committed:
Only run on files in `generated/` directory

kckeiks committed:
Resolve conflict

kckeiks committed:
Remove prost from core

kckeiks committed:
Fix core tests

kckeiks committed:
Fmt code

kckeiks committed:
Merge remote-tracking branch 'origin/add-quickprotobuf-take2' into add-quickprotobuf-take2

kckeiks committed:
Include generated in proto mod

kckeiks committed:
Remove prost from autonat

kckeiks committed:
Update changelog

thomaseizinger commented on core/src/lib.rs:
```diff
@@ -37,18 +37,12 @@
-#[allow(clippy::derive_partial_eq_without_eq)]
-mod peer_record_proto {
-    include!(concat!(env!("OUT_DIR"), "/peer_record_proto.rs"));
+mod proto {
+    include!("generated/mod.rs");
+    pub use self::{
+        envelope_proto::*, keys_proto::*, peer_record_proto::mod_PeerRecord::*,
+        peer_record_proto::PeerRecord,
+    };
 }
```
:+1:

thomaseizinger commented on core/src/lib.rs:
```diff
@@ -37,18 +37,12 @@
-mod peer_record_proto {
-    include!(concat!(env!("OUT_DIR"), "/peer_record_proto.rs"));
+mod proto {
+    include!("generated/mod.rs");
```
Why do we need an `include` here? Wouldn't it be simpler to just declare the `generated` module as a normal module on the root level and then just re-export?

thomaseizinger replied:
Perhaps the `--single-mod` option helps us removing some of this boilerplate?

kckeiks replied:
My concern is devs importing from root or proto inconsistently. I can add it to root though, no problem.

thomaseizinger replied:
Ah that is a valid concern! Happy to stick to `include` then! I am quite surprised this works though. This just imports the content of the file right? Why doesn't rustc expect those modules to be within a `proto` directory when the modules are declared within a `proto` module?

kckeiks replied:
I think it's because include! parses the file as an expression.

thomaseizinger commented on protocols/autonat/src/protocol.rs:
```diff
@@ -150,28 +154,30 @@ impl DialRequest {
+        let msg = proto::Message {
+            type_pb: Some(proto::MessageType::DIAL),
+            dial: Some(proto::Dial {
+                peer: Some(proto::PeerInfo {
+                    id: Some(Cow::from(peer_id)),
                     addrs,
                 }),
             }),
-            dial_response: None,
+            dialResponse: None,
```
This is odd. Does `pb-rs` not generate idiomatic Rust identifiers?

kckeiks replied:
No it doesn't :/

thomaseizinger reviewed (commented):
Stellar work, very much looking forward to this! Does this encompass all crates already that depend on `prost` currently? (Modulo the ones using `prost-codec`?

kckeiks committed:
Merge branch 'master' into add-quickprotobuf-take2

kckeiks committed:
Rename to quick-protobuf-codec

kckeiks committed:
Remove prost-codec

thomaseizinger commented on misc/quick-protobuf-codec/src/lib.rs:
```diff
@@ -31,21 +30,22 @@ impl<In, Out> Codec<In, Out> {
 }
 
-impl<In, Out: Message + Default> Decoder for Codec<In, Out> {
+impl<In, Out: for<'a> MessageRead<'a> + Default> Decoder for Codec<In, Out> {
```
Have you tried to fix this to `'static`?

kckeiks replied:
That doesn't work because `msg` in `Self::Item::from_reader(&mut reader, &msg)` does not live long enough.

thomaseizinger reviewed (commented):
One idea, I'll play around with this some time tomorrow.

thomaseizinger committed:
Don't use `Cow`

kckeiks committed:
Fix core

kckeiks committed:
Run fmt

kckeiks committed:
Remove prost from identify

kckeiks committed:
Fix autonat

kckeiks committed:
Update changelogs

kckeiks committed:
Update docs

kckeiks committed:
Remove prost from floodsub

kckeiks committed:
Remove prost from gossipsub

kckeiks committed:
Update changelog in identify

kckeiks committed:
Remove prost from kad

kckeiks committed:
Clean up relay

kckeiks committed:
Remove prost from rendezvous

kckeiks committed:
Remove prost from noise

kckeiks committed:
Remove prost from plaintext

kckeiks committed:
Remove prost from webrtc

kckeiks committed:
Clean up

kckeiks committed:
Fix cargo check with all features

kckeiks committed:
Fix webrtc tests

kckeiks committed:
Fix clippy errors

kckeiks committed:
Fix clippy errors take 2

kckeiks committed:
Fix CI

kckeiks committed:
Merge branch 'master' into add-quickprotobuf-take2

thomaseizinger commented on core/src/identity.rs:
```diff
@@ -162,35 +163,37 @@ impl Keypair {
         };
 
-        Ok(pk.encode_to_vec())
+        let mut buf = Vec::with_capacity(pk.get_size());
+        let mut writer = Writer::new(&mut buf);
+        pk.write_message(&mut writer).expect("Encoding to succeed");
```
Can we use `serialize_into_vec` here?
 https://docs.rs/quick-protobuf/0.8.1/quick_protobuf/writer/fn.serialize_into_vec.html

thomaseizinger replied:
And in other places?

kckeiks replied:
I thought we agreed not to use `serialize_into_vec/slice`? https://github.com/libp2p/rust-libp2p/pull/3066#discussion_r1049193826

thomaseizinger replied:
Ah right, I forgot that this writes the length as well.

kckeiks committed:
Merge branch 'master' into add-quickprotobuf-take2

thomaseizinger commented on misc/quick-protobuf-codec/CHANGELOG.md:
```diff
@@ -1,3 +1,9 @@
+# 0.1.0 [unreleased]
+
+- Migrate from `prost` to `quick-protobuf`. This removes `protoc` dependency. See [PR 3312].
+
+[PR 3312]: https://github.com/libp2p/rust-libp2p/pull/3312
```
Can you delete the old changelog entries below?

thomaseizinger commented on .github/workflows/ci.yml:
```diff
@@ -292,3 +292,30 @@ jobs:
+            **/generated/*.proto
+
+      - name: Generate proto files
+        run: pb-rs -D ${{ steps.glob.outputs.paths }}
```
Can we use the long option name here please? That would be self-descriptive then.

thomaseizinger replied:
An optimization idea for the future: We could use `Cow` in several places actually, just not all of them.

We would have to find a way how we can declare here, which `.proto` files should be generated with `-D` and which ones can omit it.

thomaseizinger commented on protocols/autonat/src/protocol.rs:
```diff
@@ -184,29 +190,27 @@ pub enum ResponseError {
 }
 
-impl From<ResponseError> for i32 {
+impl From<ResponseError> for proto::ResponseStatus {
```
Is `ResponseError` exposed in the public API? If not, we could omit it now for the generated type.

kckeiks replied:
Yes

thomaseizinger commented on protocols/dcutr/CHANGELOG.md:
```diff
@@ -11,10 +11,14 @@
 [issue 2217]: https://github.com/libp2p/rust-libp2p/issues/2217
 [PR 3214]: https://github.com/libp2p/rust-libp2p/pull/3214
+[PR 3312]: https://github.com/libp2p/rust-libp2p/pull/3312
+
```
```suggestion
```

No need for this newline.

thomaseizinger commented on .github/workflows/ci.yml:
```diff
@@ -292,3 +292,30 @@ jobs:
+
+      - uses: Swatinem/rust-cache@359a70e43a0bb8a13953b04a90f76428b4959bb6 # v2.2.0
+
+      - run: cargo install pb-rs --locked
```
Can we pin the version here please to avoid new releases breaking our CI?

thomaseizinger commented on protocols/dcutr/src/protocol/inbound.rs:
```diff
@@ -128,7 +125,7 @@ impl PendingConnect {
 pub enum UpgradeError {
     #[error(transparent)]
-    Codec(#[from] prost_codec::Error),
+    Codec(#[from] quick_protobuf_codec::Error),
```
Is `UpgradeError` part of the public API? If yes, can you create a `CodecError` newtype please like we do in `libp2p-floodsub` to hide `quick-protobuf-codec` from the public API?

kckeiks replied:
Looks like it is https://github.com/libp2p/rust-libp2p/pull/3214/files#diff-1139510a64e6bc16c649cc904d0c3c799663ee17f7828f2dea6839cfb3f5c283R39.

I see that there are several places where this pattern has been implemented. I suggest we split this into a separate task because it's a chunk of work and I fear I'm gonna have limited bandwidth soon.

thomaseizinger commented on protocols/floodsub/CHANGELOG.md:
```diff
@@ -6,7 +6,10 @@
+- Migrate from `prost` to `quick-protobuf`. This removes `protoc` dependency. See [PR 3312].
+
 [pr 3224]: https://github.com/libp2p/rust-libp2p/pull/3224
+[PR 3312]: https://github.com/libp2p/rust-libp2p/pull/3312
```
As we can see here, we only migrated to `prost-codec` recently which hasn't been released yet.

Thus, the previous changelog entry is meaningless to our users.

Can you merge the two entries please?

thomaseizinger replied:
The same may apply to other crates, please check all changelogs!

thomaseizinger reviewed (commented):
This looks really good, thanks for this work!

I've left a few minor comments but I am excited to merge this soon :)

kckeiks committed:
Merge branch 'master' into add-quickprotobuf-take2

kckeiks committed:
Fix pb-rs version

kckeiks committed:
Use long option name

kckeiks committed:
Address changelog feedback

kckeiks committed:
Remove line

kckeiks committed:
Merge branch 'add-quickprotobuf-take2' of github.com:kckeiks/rust-libp2p into add-quickprotobuf-take2

thomaseizinger commented on protocols/gossipsub/src/behaviour.rs:
```diff
@@ -178,8 +178,8 @@ impl From<MessageAuthenticity> for PublishConfig {
                 let key = if key_enc.len() <= 42 {
-                    // The public key can be inlined in [`rpc_proto::Message::from`], so we don't include it
-                    // specifically in the [`rpc_proto::Message::key`] field.
+                    // The public key can be inlined in [`rpc_proto::proto::::Message::from`], so we don't include it
+                    // specifically in the [`rpc_proto::proto::Message::key`] field.
```
```suggestion
                    // The public key can be inlined in [`proto::Message::from`], so we don't include it
                    // specifically in the [`proto::Message::key`] field.
```

thomaseizinger reviewed (dismissed):
Nice work!

I think this is ready to merge. I'll give @mxinden a few days to have a look too!

mxinden commented on core/src/identity.rs:
```diff
@@ -162,35 +163,37 @@ impl Keypair {
-        self.r#type.zeroize();
-        self.data.zeroize();
+        self.Type = proto::KeyType::default();
+        self.Data = Vec::default();
```
I don't think this properly zeroizes the underlying memory region, right? Can you not call `zeroize` on `Type` and `Data`?

kckeiks replied:
`self.r#type` is i32 but `self.Type` is an enum and it doesn't satisfy the `DefaultIsZeroes` bound necessary to be zeroizable. I'll update `Data` though.

mxinden commented on .github/workflows/ci.yml:
```diff
@@ -292,3 +292,30 @@ jobs:
+            **/generated/*.proto
+
+      - name: Generate proto files
+        run: pb-rs --dont_use_cow ${{ steps.glob.outputs.paths }}
```
What is the reasoning for not using `Cow`? I am assuming code simplicity? I find it a neat feature, though I would assume that the performance gain is not worth the additional complexity in most cases. Can we easily use `Cow` on a case-by-case basis in the future?

thomaseizinger replied:
The problem / feature is that when using `Cow`, the deserialised data structure retains a reference to the byte buffer.

This doesn't play well with `asynchronous-codec` where `Encoder` passes you a byte buffer and expects you to create an owned type, without references to the original buffer. 

I think `Encoder` would have to undergo some serious refactoring to make this possible. Perhaps now that GATs are stable, this is actually a viable interface. For the time being, we opted to not go down this route.

thomaseizinger replied:
> What is the reasoning for not using `Cow`? I am assuming code simplicity? I find it a neat feature, though I would assume that the performance gain is not worth the additional complexity in most cases. Can we easily use `Cow` on a case-by-case basis in the future?

We would be able to use `Cow` on a case-by-case basis if we find a way to run this CI check on all proto files, without having to enumerate them all so we don't forget adding new ones but also have a list of files that we want to enable `Cow` for.

mxinden commented on protocols/autonat/src/protocol.rs:
```diff
@@ -269,35 +270,38 @@ impl DialResponse {
-        bytes
+        let mut buf = Vec::with_capacity(msg.get_size());
+        let mut writer = Writer::new(&mut buf);
+        msg.write_message(&mut writer).expect("Encoding to succeed");
```
Nit pick, I would prefer the wording `.expect("Vec<u8> provides capacity as needed")`. Yes, we expect the encoding to succeed, but why do we do so? Because we only expect errors on memory allocation and given that we are using a `Vec` which can dynamically allocate, this is not an issue. Or am I misunderstanding the reasoning?

thomaseizinger replied:
You are completely on point. It think it would be useful to upstream the following to `quick-protobuf`:


```rust
/// A trait to handle deserialization based on parsed `Field`s
pub trait MessageWrite: Sized {

	// Other functions as currently defined ...

	fn write_to_vec(&self) -> Vec<u8> {
		let mut vec = Vec::with_capacity(self.get_size());
		let mut writer = Writer::new(&mut vec);

		self.write_message(&mut writer).expect("Vec<u8> provides enough capacity");

		vec
	}
}
```

On this note, I think there is some clean-up to be done in terms of APIs. It is quite confusing that there is `MessageWrite::write_message` which writes to a `Writer` and there is also `Writer::write_message` which writes the message using `MessageWriter::write_message` **with a length-prefix**!

thomaseizinger replied:
Opened an issue here: https://github.com/tafia/quick-protobuf/issues/246

kckeiks replied:
> Nit pick, I would prefer the wording .expect("Vec<u8> provides capacity as needed"). Yes, we expect the encoding to succeed, but why do we do so? Because we only expect errors on memory allocation and given that we are using a Vec which can dynamically allocate, this is not an issue. Or am I misunderstanding the reasoning?

It wasn't clear from the documentation and code in what cases we should expect errors https://github.com/libp2p/rust-libp2p/pull/3066#discussion_r1006239290 so that is why the message is so vague. Do we want this changed in this PR or wait for changes upstream?

mxinden replied:
I am fine merging as is.

mxinden reviewed (commented):
Thanks for all the work here. 3 questions. Otherwise good to go from my end.

kckeiks committed:
Merge branch 'master' into add-quickprotobuf-take2

kckeiks committed:
Zeroize Data

kckeiks committed:
Remove protoc installation from ci

kckeiks committed:
Remove protoc installation from interop-tests

kckeiks committed:
Revert ci changes

kckeiks committed:
Fmt relay

kckeiks committed:
Merge branch 'master' into add-quickprotobuf-take2

kckeiks committed:
Remove transparent

kckeiks committed:
Add protoc installation to ci

kckeiks committed:
Remove display impl

kckeiks committed:
Box error

kckeiks committed:
Change DecodeError

thomaseizinger commented on core/src/lib.rs:
```diff
@@ -75,6 +74,44 @@ pub use translation::address_translation;
 
-#[derive(thiserror::Error, Debug)]
-#[error(transparent)]
-pub struct DecodeError(Box<dyn Error + Send + Sync + 'static>);
```
I am inclined to leave this as, accept that we are removing the `UnwindSafe` auto trait and just do another major version bump. I don't think complicating the API with an additional enum is worth it at this stage.

@mxinden What do you think?

thomaseizinger replied:
Actually, lets do the following:

- Make the enum private and wrap it in a new type that doesn't expose anything.
- Record an issue to replace the enum with `quick_protobuf::Error`
- Track the issue in the `0.52` milestone
- Optional: Prepare a PR that fixes the above issue by removing the enum again and add that to `0.52` as well.

That way, we can merge this as a non-breaking change and merge in the breaking change as soon as we decide to do another breaking change. I think very likely, nobody actually depends on this error being `UnwindSafe`. I don't actually care about removing this auto trait but I do care about not making as many major version bumps.

kckeiks replied:
> Make the enum private and wrap it in a new type that doesn't expose anything.

Since the new type doesn't expose anything, I removed the enum.

I created https://github.com/libp2p/rust-libp2p/issues/3534 but can't add issues to milestones.

thomaseizinger replied:
Thank you, I've added it to the milestone!

kckeiks committed:
Merge branch 'master' into add-quickprotobuf-take2

kckeiks committed:
Make an error type that does not expose anything

thomaseizinger committed:
Merge branch 'master' into add-quickprotobuf-take2

mxinden committed:
Merge branch 'master' into add-quickprotobuf-take2

mxinden committed:
Merge branch 'master' into add-quickprotobuf-take2
